bzl/plaidml.bzl

```diff
@@ -141,3 +154,220 @@ plaidml_settings = rule(
+)
+
+
+def _plaidml_device_test_builder_impl(ctx):
+    ctx.actions.expand_template(
+        template = ctx.file._tpl,
+        output = ctx.outputs.executable,
+        substitutions = {},
+        is_executable = True
+    )
+
+    runfiles = ctx.runfiles()
+    for test in ctx.attr.tests:
+        runfiles = runfiles.merge(test[DefaultInfo].default_runfiles)
+
+    return [DefaultInfo(runfiles = runfiles, executable = ctx.outputs.executable)]
+
+
+# A rule to create a device-specific test builder script, used as a
+# tool for packaging device-specific tests for use outside of the
+# Bazel environment.
+#
+# Note that the device-specific tests are built using a transition to
+# an explicit //plaidml:target configuration; the command-line and
+# default //plaidml:target setting are *not* used.  This makes it
+# possible to build multiple device-specific tests via a single bazel
+# invocation.
+_plaidml_device_test_builder = rule(
+    attrs = {
+        "plaidml_target": attr.string(mandatory = True),
+        "tests": attr.label_list(allow_empty = False, cfg = plaidml_target),
+        "_tpl": attr.label(default = "//bzl:device_test_builder.tpl.py", allow_single_file = True),
+
+        # Allow this rule to use transitions (which bazel is fairly
+        # careful about, since transitions have the potential to
+        # explode the build graph).
+        #
+        # TODO: This has been renamed from "whitelist" to "allowlist"
+        # in more recent versions of bazel.
+        "_whitelist_function_transition": attr.label(
+            default = "@bazel_tools//tools/whitelists/function_transition_whitelist"
+        ),
+    },
+    outputs = {
+        "executable": "%{name}.py",
+    },
+    implementation = _plaidml_device_test_builder_impl
+)
+
+
+def _plaidml_device_test_runner_impl(ctx):
+    # Build the test runner.
+    script = [
+        "#!/usr/bin/env python",
+        "",
+        "import os",
+        "from pathlib import Path",
+        "import subprocess",
+        "",
+        "p = Path(__file__).parents[0]",
+        "os.chdir(p)",
+        "os.environ['RUNFILES_DIR'] = str(p)",
+        "os.environ['PLAIDML_TARGET'] = '{}'".format(ctx.attr.plaidml_target),
+        ""
+    ]
+
+    for test, args in zip(ctx.attr.tests, [arg[_ArgInfo].args for arg in ctx.attr.args]):
+        workspace = test.label.workspace_name
+        if not workspace:
+            workspace = ctx.workspace_name
+        command = [test.files_to_run.executable.short_path] + args
+        script.append("subprocess.check_call(['{}'], cwd='{}')".format("', '".join(command), workspace))
+
+    ctx.actions.write(ctx.outputs.executable, "\n".join(script), is_executable=True)
+
+    return [DefaultInfo(executable = ctx.outputs.executable)]
+
+
+# A rule to create a device-specific test runner script, which is
+# packaged into an out-of-bazel device-specific unit testing package.
+#
+# Note that the device-specific tests are built using a transition to
```

Comment:
When you say "device-specific", do you actually mean "target-specific". I don't see anything in here that specifies the device, all I see is mentioning the target. So I guess I'm a bit confused.

Reply:
Sorry -- yes, target-specific.  I'll reword for clarity.

Reply:
Great! Thanks, that makes a ton more sense to me now.